### PR TITLE
Add abutting directions to excision spheres

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

This is the information needed to figure out which elements face the excision boundary. Used for interpolations from element faces to a Strahlkorper representing the excision.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

**Comment by nilsvu** [comment]
@gsb76 @markscheel I needed this block neighbor information to get observations on excision boundaries to an (almost!) working state, and I remember we discussed adding this a while back. I hope this helps you as well.

**Comment by nilsvu** [comment]
@nikwit ping because I think you are using something like this on a branch as well

**Comment by wthrowe** [comment]
Why have you chosen a set here?  It looks like you are using it as a map.  (Iterating over all the pairs looking for a particular first entry.)

**Comment by nilsvu** [comment]
I think you're absolutely right that a map makes more sense. Give me a sec to check.

File: src/Domain/Structure/ExcisionSphere.hpp [hunk 1]
```diff
@@ -27,14 +33,21 @@ class er;
...
+
  public:
   /// Constructor
   ///
   /// \param radius the radius of the excision sphere in the
   /// computational domain.
   /// \param center the coordinate center of the excision sphere
   /// in the computational domain.
-  ExcisionSphere(double radius, std::array<double, VolumeDim> center);
+  /// \param block_neighbors the set of blocks that face the excision sphere.
```
Comment (kidder) [inline src/Domain/Structure/ExcisionSphere.hpp]: face ->  abut

face just means oriented towards, while abut means touching

File: src/Domain/Structure/ExcisionSphere.hpp [hunk 2]
```diff
@@ -52,13 +65,30 @@ class ExcisionSphere {
...
   const std::array<double, VolumeDim>& center() const { return center_; }
 
+  /// The set of blocks that face the excision sphere. Each element in this set
+  /// is a pair (block ID, face direction).
+  const auto& block_neighbors() const { return block_neighbors_; }
+
+  /// The direction in which the block identified by `block_id` faces the
+  /// excision sphere, or `std::nullopt` if the block does not face the excision
+  /// sphere
+  std::optional<Direction<VolumeDim>> face_direction(
```
Comment (kidder) [inline src/Domain/Structure/ExcisionSphere.hpp]: face_direction -> abutting_direction

File: src/Domain/Creators/BinaryCompactObject.cpp [hunk 3]
```diff
@@ -511,10 +511,22 @@ Domain<3> BinaryCompactObject::create_domain() const {
       std::move(boundary_conditions_all_blocks),
       {{"ObjectAExcisionSphere",
         ExcisionSphere<3>{object_A_.inner_radius,
-                          {{object_A_.x_coord, 0.0, 0.0}}}},
+                          {{object_A_.x_coord, 0.0, 0.0}},
```
Comment (mar-celine) [inline src/Domain/Creators/BinaryCompactObject.cpp:527]: Maybe a small comment saying "Direction of faces given by [reference to either documentation or code]"

**Review by kidder — CHANGES_REQUESTED** [review]

**Review by mar-celine — COMMENTED** [review]
Do you know if there's any need to also keep track of the orientation/connectivity of the faces? All of excision boundary directions being lower zeta doesn't determine this and I wonder if knowing which excision boundary faces are neighboring each other is information you might need. Otherwise looks good!

**Comment by nilsvu** [comment]
I simplified this a lot since @wthrowe is absolutely right that a map is the right representation here. Thanks!

Also squashed in @kidder's and @gsb76's comments.

@gsb76 I believe we don't need orientations and connectivity. My current thought is: a surface integral over the excision boundary decomposes into the sum of the integrals over the individual abutting element faces. So to get the spherical harmonic coefficients we can integrate the spherical harmonics times the data over each abutting element face independently and sum the result. This can be expressed as a projection matrix I believe, but should only need information about the element face. Either way, if I'm wrong we can add the orientation information when we need it.

**Review by mar-celine — DISMISSED** [review]
Looks good! :+1:

**Review by wthrowe — APPROVED** [review]

**Comment by nilsvu** [comment]
@kidder please check if your comments have been addressed here.

**Review by kidder — APPROVED** [review]

**Pull request merged**